rfc/20250211-s3-locking-with-conditional-writes.md

```diff
@@ -0,0 +1,148 @@
+    * The generation of the MD5 checksum is already part of the [implementation](https://github.com/opentofu/opentofu/blob/6614782/internal/backend/remote-state/s3/client.go#L178).
+    * We just need to store it in a new file in the configured S3 bucket (file name: <state-file-path>-md5)
+* When the new attribute `use_lockfile` will exist **alongside** `dynamodb_table`, OpenTofu will:
+  * Acquire the lock in the S3 bucket;
+  * Get the digest of the statefile from S3. If missing:
+    * Acquire the lock in DynamoDB table; 
+    * Get the digest of the statefile from DynamoDB and migrate it to the S3 bucket;
+    * Release the lock from DynamoDB table;
+  * Perform the requested sub-command
+  * Release the lock from the S3 bucket;
+
+> [!NOTE]
+>
+> OpenTofu [recommends](https://opentofu.org/docs/language/settings/backends/s3/) to have versioning enabled for the S3 buckets used to store state files.
+>
+> Acquiring and releasing locks will add a good amount of writes and reads to the bucket. Therefore, for a versioning-enabled S3 bucket, the number of versions for that file could grow significantly.
+> Even though the cost should be negligible for the locking files, any user using this feature could consider configuring the lifecycle of the S3 bucket to limit the number of versions of a file.
+
+### Technical Approach
+
+In order to achieve and ensure a proper state locking via S3 bucket, we want to attempt to create the locking object only when it is missing. 
+In order to do so we need to call `s3client.PutObject` with the property `IfNoneMatch: "*"`.
+For more information, please check the [official documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html#conditional-write-key-names).
+
+But the simplified implementation would look like this:
+```go
+input := &s3.PutObjectInput{
+    Bucket:            aws.String(bucket),
+    Key:               aws.String(key),
+    Body:              bytes.NewReader([]byte(lockInfo)),
+    IfNoneMatch:       aws.String("*"),
+}
+_, err := actor.S3Client.PutObject(ctx, input)
+```
+
+The `err` returned above should be handled accordingly with the [behaviour defined](https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html) in the official docs:
+* HTTP 200 (OK) - Means that the locking file was not existing. Therefore, the lock can be considered as acquired.
+  * For buckets with versioning enabled, if there's no current object version with the same name, or if the current object version is a delete marker, the write operation succeeds.
+* HTTP 412 (Precondition Failed) - Means that the locking file is already there. Therefore, the whole process should exit because the lock couldn't be acquired.
+* HTTP 409 (Conflict) - Means that the there was a conflict of concurrent requests. AWS recommends to retry the request in such cases, but we could just handle this similarly to the 412 case.
+
+#### Digest updates
```

Comment:
I may be missing something, but do we need to update the digest? If I understand correctly, we could go with a `IfNoneMatch` lock creation so it would fail for any other tofu run, while the file exists. On Unlock, we could just delete it altogether. As far as I can tell, this matches the dynamodb implementation and eliminates data races via `IfNoneMatch` option. 

What do you think?

Reply:
I used terms like "digest", "md5" or "digest file" for the part where OpenTofu is writing an md5 hash of the statefile.
For example, looking into the dynamoDB, there is the md5 sum of the state file. Can be easily checked like this:
![image](https://github.com/user-attachments/assets/7dc94a57-407b-4531-b6db-02bc2e24d10b)

Reply:
Got you, thanks! I missed that part, I will re-read it and leave another comment.

Reply:
Updated this section due to not needed to handle this digest file anymore.
This should be handled only when the DynamoDB locking is enabled. That's due to having two sources of information when it comes to locking and it was used to ensure that the state object from the S3 bucket was not altered separately, without acquiring a lock.

Reply:
After some digging, I think this option is unnecessary even for dynamodb locking. This feature was introduced to workaround the eventual consistency nature of s3 (before 2020). It is checking if the read state file is the latest written one and we no longer need to check this since s3 become strongly consistent ([blogpost](https://aws.amazon.com/blogs/aws/amazon-s3-update-strong-read-after-write-consistency/)) in 2020.

Reply:
This is something only slightly relevant for this RFC, so I think we can tackle it separately. Alternatively, we can leave it as is if we plan to deprecate dynamodb locking completely. Otherwise, removing md5 write / read from dynamodb locking will positively affect the performance.

Reply:
Added in 38f1eb921775d9ac783fffe8c8cb3d2b11a67419